File: encryption_zone/encrypt.py
import os
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives import padding
import secrets

# Function to generate a random encryption key
def generate_key():
    return secrets.token_bytes(32)  # AES-256

# Function to encrypt a file
def encrypt_file(file_path, key):
    try:
        print(f"Encrypting: {file_path}")  # Print the file name before encryption

        # Open the file in binary mode
        with open(file_path, 'rb') as file:
            file_data = file.read()

        if not file_data:
            print(f"File is empty or could not be read: {file_path}")
            return

        # Pad the data to be multiple of block size (128 bits for AES)
        padder = padding.PKCS7(128).padder()
        padded_data = padder.update(file_data) + padder.finalize()

        # Create a random IV for each encryption
        iv = secrets.token_bytes(16)

        # Set up the AES cipher in CBC mode
        cipher = Cipher(algorithms.AES(key), modes.CBC(iv), backend=default_backend())
        encryptor = cipher.encryptor()

        # Encrypt the data
        encrypted_data = encryptor.update(padded_data) + encryptor.finalize()

        # Write the encrypted data back to the file (overwrite the original or save as new)
        with open(file_path + '.enc', 'wb') as encrypted_file:
            encrypted_file.write(iv + encrypted_data)  # Prepend the IV to the encrypted file

        print(f"Encrypted: {file_path}")
    except Exception as e:
        print(f"Failed to encrypt {file_path}: {e}")

# Function to encrypt all files in a folder and subfolders
def encrypt_folder(folder_path, key):
    if not os.path.exists(folder_path):
        print(f"Folder does not exist: {folder_path}")
        return

    print(f"Starting encryption in folder: {folder_path}")
    # Walk through the directory
    for dirpath, dirnames, filenames in os.walk(folder_path):
        print(f"Processing directory: {dirpath}")
        for filename in filenames:
            file_path = os.path.join(dirpath, filename)
            print(f"Found file: {file_path}")
            encrypt_file(file_path, key)

# Main function
if __name__ == "__main__":
    folder_path = "./encryption_zone/files2encrypt"  # Always use the "files2encrypt" folder
    key = generate_key()
    encrypt_folder(folder_path, key)
    print("All files encrypted.")
    print("Encryption key:", key.hex())  # Convert bytes to hex string for printing

File: encryption_zone/encrypt.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <filesystem>
#include <openssl/aes.h>
#include <openssl/rand.h>
#include <openssl/pem.h>
#include <openssl/evp.h>

using namespace std;
namespace fs = std::filesystem;

// Function to generate a random encryption key
vector<unsigned char> generate_key() {
    vector<unsigned char> key(32); // AES-256
    if (!RAND_bytes(key.data(), key.size())) {
        throw runtime_error("Failed to generate key");
    }
    return key;
}

// Function to encrypt a file
void encrypt_file(const string& file_path, const vector<unsigned char>& key) {
    try {
        cout << "Encrypting: " << file_path << endl; // Print the file name before encryption

        // Open the file in binary mode
        ifstream file(file_path, ios::binary);
        if (!file) {
            throw runtime_error("File could not be opened");
        }

        vector<unsigned char> file_data((istreambuf_iterator<char>(file)), istreambuf_iterator<char>());
        file.close();

        if (file_data.empty()) {
            cout << "File is empty or could not be read: " << file_path << endl;
            return;
        }

        // Pad the data to be multiple of block size (128 bits for AES)
        int block_size = AES_BLOCK_SIZE;
        int padding_length = block_size - (file_data.size() % block_size);
        file_data.insert(file_data.end(), padding_length, static_cast<unsigned char>(padding_length));

        // Create a random IV for each encryption
        vector<unsigned char> iv(AES_BLOCK_SIZE);
        if (!RAND_bytes(iv.data(), iv.size())) {
            throw runtime_error("Failed to generate IV");
        }

        // Set up the AES cipher in CBC mode
        EVP_CIPHER_CTX* ctx = EVP_CIPHER_CTX_new();
        EVP_EncryptInit_ex(ctx, EVP_aes_256_cbc(), nullptr, key.data(), iv.data());

        vector<unsigned char> encrypted_data(file_data.size() + AES_BLOCK_SIZE);
        int out_len1, out_len2;
        EVP_EncryptUpdate(ctx, encrypted_data.data(), &out_len1, file_data.data(), file_data.size());
        EVP_EncryptFinal_ex(ctx, encrypted_data.data() + out_len1, &out_len2);
        EVP_CIPHER_CTX_free(ctx);

        // Write the encrypted data back to the file (overwrite the original or save as new)
        ofstream encrypted_file(file_path + ".enc", ios::binary);
        encrypted_file.write(reinterpret_cast<char*>(iv.data()), iv.size()); // Prepend the IV to the encrypted file
        encrypted_file.write(reinterpret_cast<char*>(encrypted_data.data()), out_len1 + out_len2);
        encrypted_file.close();

        cout << "Encrypted: " << file_path << endl;
    } catch (const exception& e) {
        cout << "Failed to encrypt " << file_path << ": " << e.what() << endl;
    }
}

// Function to encrypt all files in a folder and subfolders
void encrypt_folder(const string& folder_path, const vector<unsigned char>& key) {
    if (!fs::exists(folder_path)) {
        cout << "Folder does not exist: " << folder_path << endl;
        return;
    }

    cout << "Starting encryption in folder: " << folder_path << endl;
    // Walk through the directory
    for (const auto& dir_entry : fs::recursive_directory_iterator(folder_path)) {
        if (fs::is_regular_file(dir_entry)) {
            string file_path = dir_entry.path().string();
            cout << "Found file: " << file_path << endl;
            encrypt_file(file_path, key);
        }
    }
}

// Main function
int main() {
    string folder_path = "./files2encrypt"; // Always use the "files2encrypt" folder
    vector<unsigned char> key = generate_key();
    encrypt_folder(folder_path, key);
    cout << "All files encrypted." << endl;

    // Convert bytes to hex string for printing
    cout << "Encryption key: ";
    for (const auto& byte : key) {
        cout << hex << setw(2) << setfill('0') << (int)byte;
    }
    cout << endl;

    return 0;
}
